File: kakao.app/js/ked/view/base/scroll.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, min: function () { var m = Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.min.apply(_k_.min,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n < m ? n : m}}}; return m }, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }, eql: function (a,b,s) { var i, k, v; s = (s != null ? s : []); if (Object.is(a,b)) { return true }; if (typeof(a) !== typeof(b)) { return false }; if (!(Array.isArray(a)) && !(typeof(a) === 'object')) { return false }; if (Array.isArray(a)) { if (a.length !== b.length) { return false }; var list = _k_.list(a); for (i = 0; i < list.length; i++) { v = list[i]; s.push(i); if (!_k_.eql(v,b[i],s)) { s.splice(0,s.length); return false }; if (_k_.empty(s)) { return false }; s.pop() } } else if (_k_.isStr(a)) { return a === b } else { if (!_k_.eql(Object.keys(a),Object.keys(b))) { return false }; for (k in a) { v = a[k]; s.push(k); if (!_k_.eql(v,b[k],s)) { s.splice(0,s.length); return false }; if (_k_.empty(s)) { return false }; s.pop() } }; return true }, isStr: function (o) {return typeof o === 'string' || o instanceof String}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}}

var pow, scroll

import kxk from "../../../kxk.js"
let post = kxk.post

import theme from "../../theme/theme.js"

import squares from "../../util/img/squares.js"
import sircels from "../../util/img/sircels.js"

import view from "./view.js"

pow = Math.pow


scroll = (function ()
{
    _k_.extend(scroll, view)
    function scroll (screen, state, side = 'left')
    {
        this.state = state
        this.side = side
    
        this["draw"] = this["draw"].bind(this)
        this["scrollToPixel"] = this["scrollToPixel"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        scroll.__super__.constructor.call(this,screen,this.state.owner() + '_scroll')
        this.pointerType = 'pointer'
        this.setColor('bg',theme.gutter.bg)
        this.setColor('dot',theme.scroll.dot)
        this.setColor('knob',theme.scroll.knob)
        this.setColor('hover',theme.scroll.hover)
    }

    scroll.prototype["onMouse"] = function (event)
    {
        var col, row

        var _a_ = this.eventPos(event); col = _a_[0]; row = _a_[1]

        scroll.__super__.onMouse.call(this,event)
        switch (event.type)
        {
            case 'press':
                if (this.hover)
                {
                    this.doDrag = true
                    post.emit('pointer','grabbing')
                    return this.scrollToPixel(event.pixel)
                }
                break
            case 'drag':
                if (this.doDrag)
                {
                    this.hover = true
                    post.emit('pointer','grab')
                    return this.scrollToPixel(event.pixel)
                }
                this.hover = false
                break
            case 'release':
                if (this.doDrag)
                {
                    if (this.hover)
                    {
                        post.emit('pointer','pointer')
                    }
                    delete this.doDrag
                    return true
                }
                break
        }

        return this.hover
    }

    scroll.prototype["isActive"] = function ()
    {
        return this.state.s.lines.length > this.cells.rows
    }

    scroll.prototype["scrollToPixel"] = function (pixel)
    {
        var csz, maxY, rowf, view

        csz = this.screen.t.cellsz
        if (_k_.empty(csz))
        {
            return
        }
        view = this.state.s.view.asMutable()
        rowf = pixel[1] / csz[1] - this.cells.y
        view[1] = Math.floor(rowf * (this.state.s.lines.length - this.cells.rows + 1) / (this.cells.rows - 1))
        maxY = this.state.s.lines.length - this.cells.rows
        if (maxY > 0)
        {
            view[1] = _k_.min(maxY,view[1])
        }
        view[1] = _k_.max(0,view[1])
        if (_k_.eql(view, this.state.s.view))
        {
            return true
        }
        this.state.setView(view)
        return {redraw:true}
    }

    scroll.prototype["draw"] = function ()
    {
        var csz, fg, h, kh, ky, lnum, rows, w, x, y

        csz = this.screen.t.cellsz
        if (_k_.empty(csz))
        {
            return
        }
        rows = this.cells.rows
        this.cells.fill_col(0,0,rows,' ',null,this.color.bg)
        lnum = this.state.s.lines.length
        if (lnum <= rows)
        {
            return
        }
        kh = ((rows * rows) / lnum) * csz[1]
        ky = ((rows * csz[1] - kh) * this.state.s.view[1] / (lnum - rows))
        fg = (this.hover ? this.color.hover : this.color.knob)
        x = this.cells.x * csz[0]
        y = parseInt(this.cells.y * csz[1] + ky)
        w = parseInt(csz[0] / 2)
        h = parseInt(kh)
        squares.place(x,parseInt(y + w / 2),w,h - w,fg)
        sircels.place(x,y,w,((ky ? fg : this.color.dot)))
        return sircels.place(x,y + h - w,w,(((y + h < (this.cells.y + rows) * csz[1] - 1) ? fg : this.color.dot)))
    }

    return scroll
})()

export default scroll;

File: kakao.app/love/lua/view/base/scroll.lua
--[[
     ███████   ███████  ████████    ███████   ███      ███      
    ███       ███       ███   ███  ███   ███  ███      ███      
    ███████   ███       ███████    ███   ███  ███      ███      
         ███  ███       ███   ███  ███   ███  ███      ███      
    ███████    ███████  ███   ███   ███████   ███████  ███████  

    scrolls a view
    used by any multiline editor
--]]

view = require "view.base.view"


local scroll = class("scroll", view)
    


function scroll:init(state, side) 
        side = side or 'left'
        
        self.state = state
        self.side = side
        view.init(self, self.state:owner() .. '_scroll')
        
        self.pointerType = 'pointer'
        
        self:setColor('bg', theme.gutter.bg)
        self:setColor('dot', theme.scroll.dot)
        self:setColor('knob', theme.scroll.knob)
        self:setColor('hover', theme.scroll.hover)
        return self
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function scroll:onMouse(event) 
        local col, row = self:eventPos(event)
        
        view.onMouse(self, event)
        
        if (event.type == 'press') then 
                if self.hover then 
                    self.doDrag = true
                    post:emit('pointer', 'grabbing')
                    return self:scrollToPixel(event.pixel)
                end
        elseif (event.type == 'drag') then 
                if self.doDrag then 
                    self.hover = true
                    post:emit('pointer', 'grab')
                    return self:scrollToPixel(event.pixel)
                end
                
                self.hover = false
        elseif (event.type == 'release') then 
                if self.doDrag then 
                    if self.hover then 
                        post:emit('pointer', 'pointer')
                    end
                    
                    self.doDrag = nil
                    return true
                end
        end
        
        return self.hover
    end


function scroll:isActive() 
        return (#self.state.s.lines > self.cells.rows)
    end

--  0000000   0000000  00000000    0000000   000      000      000000000   0000000   
-- 000       000       000   000  000   000  000      000         000     000   000  
-- 0000000   000       0000000    000   000  000      000         000     000   000  
--      000  000       000   000  000   000  000      000         000     000   000  
-- 0000000    0000000  000   000   0000000   0000000  0000000     000      0000000   


function scroll:scrollToPixel(pixel) 
        local csz = self.screen:size()
        
        local view = self.state.s.view:arr()
        
        local rowf = ((pixel[2] / csz[2]) - self.cells.y)
        view[2] = floor(((rowf * ((self.state.s.lines:len() - self.cells.rows) + 1)) / (self.cells.rows - 1)))
        
        local maxY = (self.state.s.lines:len() - self.cells.rows)
        
        if (maxY > 1) then 
            view[2] = min(maxY, view[2])
        end
        
        view[2] = max(1, view[2])
        
        if view:eql(self.state.s.view) then return true end
        
        self.state:setView(view)
        
        return true
    end

-- 0000000    00000000    0000000   000   000  
-- 000   000  000   000  000   000  000 0 000  
-- 000   000  0000000    000000000  000000000  
-- 000   000  000   000  000   000  000   000  
-- 0000000    000   000  000   000  00     00  


function scroll:draw() 
        local cw = _G.screen.cw
        local ch = _G.screen.ch
        
        local rows = self.cells.rows
        -- log "DRAW" @cells.cols, rows
        -- @cells∙fill_col 1 1 rows ' ' nil @color.bg
        self.cells:fill_col(1, 1, rows, 'x', array(55, 0, 0), array(255, 0, 0))
        
        -- lnum = @state.s.lines.len
        -- 
        -- ⮐  if lnum <= rows
        -- 
        -- kh = ((rows*rows) / lnum) * ch
        -- ky = ((rows*ch-kh) * @state.s.view[2] / (lnum-rows)) 
        
        -- fg = if @hover ➜ @color.hover ➜ @color.knob
        
        -- x  = @cells.x*cw
        -- y  = int @cells.y*ch+ky
        -- w  = int cw/2
        -- h  = int kh
        
        -- squares∙place x int(y+w/2) w h-w fg
        -- 
        -- sircels∙place x y     w (ky or {fg: @color.dot}) 1111
        -- sircels∙place x y+h-w w ((y+h < (@cells.y+rows)*cw-1) or {fg: @color.dot}) 1111
        
        return self:render()
    end

return scroll